Regression of issue #5015

It seems in early August during the cleaning up of ./dependencies/llvm/CMakeLists.txt, the fixes for issue #5015 seem to have been lost.  I was able to get it working with a 3 line patch, but I suspect that this isn't the fix you are ultimately going to want.

```diff
diff --git a/dependencies/llvm/CMakeLists.txt b/dependencies/llvm/CMakeLists.txt
index 7222f0582..4d13f1ac4 100644
--- a/dependencies/llvm/CMakeLists.txt
+++ b/dependencies/llvm/CMakeLists.txt
@@ -71,7 +71,10 @@ endforeach ()
 option(Halide_BUNDLE_LLVM "When built as a static library, include LLVM's objects." OFF)
 option(Halide_SHARED_LLVM "Link against shared LLVM (disables components)." OFF)
 
+if (NOT DEFINED LLVM_LIBNAMES)
 llvm_map_components_to_libnames(LLVM_LIBNAMES ${LLVM_COMPONENTS})
+endif ()
+
 if (TARGET_WEBASSEMBLY)
     find_package(LLD CONFIG REQUIRED HINTS "${LLVM_DIR}/../lld")
     list(APPEND LLVM_LIBNAMES lldWasm)
```

I only have x86 and NVPTX backends on my llvm install, so I couldn't test all of the backends.  With this patch, I was able to build Halide when I passed the following arguments to CMake

```console
cmake -DHalide_BUNDLE_LLVM=OFF -DHalide_SHARED_LLVM=ON -DBUILD_SHARED_LIBS=ON .. -DLLVM_LIBNAMES=$(llvm-config --libs) -G Ninja
```

@alexreinking I use gentoo linux which uses slotted LLVM installation which allows multiple simultaneous installations of LLVM into different directories similar to what an HPC system would do with modules.  Gentoo only build shared libraries by default, and doesn't have a way to build static libraries easily for the LLVM packages.

The slotted  install looks kind of like this

```console
[runderwood@voyager build]$ tree /usr/lib/llvm -L 2
/usr/lib/llvm
├── 10
│   ├── bin
│   ├── include
│   ├── lib64
│   ├── libexec
│   └── share
└── 9
    ├── bin
    ├── include
    ├── lib64
    ├── libexec
    └── share

```

llvm-config outputs this

```console
[runderwood@voyager ~]$ llvm-config --libs
-lLLVM-10
```

Running the build for me does this:

```console
[runderwood@voyager /tmp]$ git clone https://github.com/halide/Halide
Cloning into 'Halide'...
remote: Enumerating objects: 234, done.
remote: Counting objects: 100% (234/234), done.
remote: Compressing objects: 100% (171/171), done.
remote: Total 151731 (delta 127), reused 114 (delta 62), pack-reused 151497
Receiving objects: 100% (151731/151731), 141.51 MiB | 2.35 MiB/s, done.
Resolving deltas: 100% (113418/113418), done.
[runderwood@voyager /tmp]$ mkdir build
[runderwood@voyager /tmp]$ cd build 
[runderwood@voyager build]$ cmake -G Ninja -DCMAKE_BUILD_TYPE=Release -DHalide_SHARED_LLVM=ON ../Halide 
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Halide detected active CMake target `x86-64-linux`
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE  
-- Found OpenGL: /usr/lib64/libOpenGL.so   
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Found LLVM 10.0.1
-- Using LLVMConfig.cmake in: /usr/lib/llvm/10/lib64/cmake/llvm
-- Disabling AArch64 backend
-- Disabling AMDGPU backend
-- Disabling ARM backend
-- Disabling Hexagon backend
-- Disabling Mips backend
-- Enabling NVPTX backend
-- Disabling PowerPC backend
-- Disabling RISCV backend
-- Enabling X86 backend
CMake Error at /usr/lib/llvm/10/lib64/cmake/llvm/LLVM-Config.cmake:136 (message):
  Target NVPTX is not in the set of libraries.
Call Stack (most recent call first):
  /usr/lib/llvm/10/lib64/cmake/llvm/LLVM-Config.cmake:261 (llvm_expand_pseudo_components)
  dependencies/llvm/CMakeLists.txt:74 (llvm_map_components_to_libnames)


-- Configuring incomplete, errors occurred!
See also "/tmp/build/CMakeFiles/CMakeOutput.log".
See also "/tmp/build/CMakeFiles/CMakeError.log".

```

--- a/cmake/BundleStatic.cmake
+++ b/cmake/BundleStatic.cmake
@@ -33,16 +33,15 @@ cmake_minimum_required(VERSION 3.16)
 # IMPORTED_NO_SONAME(_<CONFIG>) # shared-only
 # IMPORTED_SONAME(_<CONFIG>) # shared-only
 
-function(bundle_static)
+function(bundle_static TARGET)
     set(options)
-    set(oneValueArgs TARGET)
+    set(oneValueArgs)
     set(multiValueArgs LIBRARIES)
     cmake_parse_arguments(ARG "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
 
-    set(interfaceLib ${ARG_TARGET})
-    set(objectLib ${ARG_TARGET}.obj)
+    set(interfaceLib ${TARGET})
+    set(objectLib ${interfaceLib}.obj)
 
-    add_library(${interfaceLib} INTERFACE)
     add_library(${objectLib} OBJECT IMPORTED)
     set_target_properties(${objectLib} PROPERTIES IMPORTED_GLOBAL TRUE)
 

--- a/dependencies/llvm/CMakeLists.txt
+++ b/dependencies/llvm/CMakeLists.txt
@@ -3,13 +3,14 @@
 ##
 
 include(CMakeDependentOption)
+include(BundleStatic)
 
 # Fallback configurations for weirdly built LLVMs
 set(CMAKE_MAP_IMPORTED_CONFIG_MINSIZEREL MinSizeRel Release RelWithDebInfo "")
 set(CMAKE_MAP_IMPORTED_CONFIG_RELWITHDEBINFO RelWithDebInfo Release MinSizeRel "")
 set(CMAKE_MAP_IMPORTED_CONFIG_RELEASE Release MinSizeRel RelWithDebInfo "")
 
-find_package(LLVM ${HALIDE_REQUIRE_LLVM_VERSION} REQUIRED)
+find_package(LLVM ${Halide_REQUIRE_LLVM_VERSION} REQUIRED)
 find_package(Clang REQUIRED CONFIG HINTS "${LLVM_DIR}/../clang")
 
 message(STATUS "Found LLVM ${LLVM_PACKAGE_VERSION}")
@@ -26,6 +27,14 @@ endif ()
 set(Halide_LLVM_DEFS ${LLVM_DEFINITIONS} $<BUILD_INTERFACE:LLVM_VERSION=${LLVM_VERSION_MAJOR}${LLVM_VERSION_MINOR}>)
 
 ##
+# Options for which version of LLVM to use
+##
+
+option(Halide_SHARED_LLVM "Link against shared LLVM (ignores components)." OFF)
+cmake_dependent_option(Halide_BUNDLE_LLVM "When built as a static library, include LLVM's objects." OFF
+                       "NOT BUILD_SHARED_LIBS" OFF)
+
+##
 # Promote LLVM/Clang executable targets
 ##
 
@@ -40,7 +49,7 @@ endif ()
 # Create options for including or excluding LLVM backends.
 ##
 
-set(LLVM_COMPONENTS mcjit bitwriter linker passes)
+set(active_components mcjit bitwriter linker passes)
 set(known_components AArch64 AMDGPU ARM Hexagon Mips NVPTX PowerPC RISCV WebAssembly X86)
 
 # We don't support LLVM10 or below for wasm codegen.
@@ -54,48 +63,48 @@ foreach (comp IN LISTS known_components)
 
     cmake_dependent_option(${OPTION} "Include ${comp} target" ON
                            "${comp} IN_LIST LLVM_TARGETS_TO_BUILD" OFF)
-    if (${OPTION})
+    if (${OPTION} OR Halide_SHARED_LLVM)
         message(STATUS "Enabling ${comp} backend")
         list(APPEND Halide_LLVM_DEFS $<BUILD_INTERFACE:${DEFINE}>)
-        list(APPEND LLVM_COMPONENTS ${comp})
+        list(APPEND active_components ${comp})
     else ()
         message(STATUS "Disabling ${comp} backend")
     endif ()
 endforeach ()
+
+set(wasm_libs "")
+if (TARGET_WEBASSEMBLY)
+    find_package(LLD CONFIG REQUIRED HINTS "${LLVM_DIR}/../lld")
+    set(wasm_libs lldWasm)
+endif ()
 
 ##
 # Create Halide::LLVM library alias pointing to the correct LLVM
 # among shared, static, and bundled.
 ##
 
-option(Halide_BUNDLE_LLVM "When built as a static library, include LLVM's objects." OFF)
-option(Halide_SHARED_LLVM "Link against shared LLVM (disables components)." OFF)
+add_library(Halide_LLVM INTERFACE)
+add_library(Halide::LLVM ALIAS Halide_LLVM)
 
-llvm_map_components_to_libnames(LLVM_LIBNAMES ${LLVM_COMPONENTS})
-if (TARGET_WEBASSEMBLY)
-    find_package(LLD CONFIG REQUIRED HINTS "${LLVM_DIR}/../lld")
-    list(APPEND LLVM_LIBNAMES lldWasm)
-endif ()
-
-if (Halide_BUNDLE_LLVM AND NOT BUILD_SHARED_LIBS)
-    include(BundleStatic)
-    bundle_static(LIBRARIES ${LLVM_LIBNAMES} TARGET Halide_LLVM)
-else ()
-    add_library(Halide_LLVM INTERFACE)
-    if (Halide_SHARED_LLVM)
-        set(LLVM_LIBNAMES LLVM)
-        target_link_libraries(Halide_LLVM INTERFACE ${CMAKE_DL_LIBS})
-    endif ()
-    target_link_libraries(Halide_LLVM INTERFACE ${LLVM_LIBNAMES})
-    set_target_properties(${LLVM_LIBNAMES} PROPERTIES IMPORTED_GLOBAL TRUE)
-endif ()
-
-# Attach the include dirs and (patched) definitions to the target, where they belong.
 set_target_properties(Halide_LLVM PROPERTIES EXPORT_NAME LLVM)
 target_include_directories(Halide_LLVM INTERFACE $<BUILD_INTERFACE:${LLVM_INCLUDE_DIRS}>)
 target_compile_definitions(Halide_LLVM INTERFACE ${Halide_LLVM_DEFS})
 
-add_library(Halide::LLVM ALIAS Halide_LLVM)
+# Link LLVM libraries to Halide_LLVM, depending on shared, static, or bundled selection.
+if (Halide_SHARED_LLVM)
+    # llvm_map_components_to_libnames is not safe to call if the LLVM static libraries
+    # aren't in the package. This happens on Gentoo Linux at least, but might also happen
+    # with custom LLVM build configurations.
+    target_link_libraries(Halide_LLVM INTERFACE LLVM ${wasm_libs} ${CMAKE_DL_LIBS})
+else ()
+    llvm_map_components_to_libnames(llvm_targets ${active_components})
+    list(APPEND llvm_targets ${wasm_libs})
+    if (Halide_BUNDLE_LLVM)
+        bundle_static(Halide_LLVM LIBRARIES ${llvm_targets})
+    else ()
+        target_link_libraries(Halide_LLVM INTERFACE ${llvm_targets})
+    endif ()
+endif ()
 
 ##
 # Language options interface library

--- a/src/Generator.cpp
+++ b/src/Generator.cpp
@@ -747,7 +747,7 @@ std::string halide_type_to_c_type(const Type &t) {
 
 int generate_filter_main_inner(int argc, char **argv, std::ostream &cerr) {
     const char kUsage[] =
-        "gengen \n"
+        "gengen\n"
         "  [-g GENERATOR_NAME] [-f FUNCTION_NAME] [-o OUTPUT_DIR] [-r RUNTIME_NAME] [-d 1|0]\n"
         "  [-e EMIT_OPTIONS] [-n FILE_BASE_NAME] [-p PLUGIN_NAME] [-s AUTOSCHEDULER_NAME]\n"
         "       target=target-string[,target-string...] [generator_arg=value [...]]\n"

